src/psyclone/transformations.py

```diff
@@ -340,67 +379,124 @@ def name(self):
+                                     `same_space` optional argument is not \
+                                     set to `True`.
+        :raises TransformationError: if the Loops are over different spaces \
+                                     that are not both discontinuous.
+        :raises TransformationError: if the Loops' upper bound names are \
+                                     not the same.
+        :raises TransformationError: if the halo-depth indices of two Loops \
+                                     are not the same.
+        :raises TransformationError: if each Loop already contains a reduction.
+        :raises TransformationError: if the first Loop has a reduction and \
+                                     the second Loop reads the result of \
+                                     the reduction.
+        :raises TransformationError: if there is an unexpected exception.
         '''
         LoopFuseTrans._validate(self, node1, node2)
 
-        from psyclone.dynamo0p3 import VALID_FUNCTION_SPACES
-        try:
+        # Set the 'same_space' flag value
+        same_space = self._same_space_flag
+
+        # Call the parent class validation first
+        super(DynamoLoopFuseTrans, self)._validate(node1, node2)
+
+        # Now test for Dynamo-specific constraints
 
-            # Check that we don't have an inter-grid kernel
+        from psyclone.dynamo0p3 import VALID_FUNCTION_SPACE_NAMES, \
+            VALID_DISCONTINUOUS_FUNCTION_SPACE_NAMES
+        try:
+            # 1) Check that we don't have an inter-grid kernel
             check_intergrid(node1)
             check_intergrid(node2)
 
-            if node1.field_space.orig_name in VALID_FUNCTION_SPACES and \
-               node2.field_space.orig_name in VALID_FUNCTION_SPACES:
-                if node1.field_space.orig_name != node2.field_space.orig_name:
-                    if same_space:
-                        info = (
-                            " Note, the same_space flag was set, but "
-                            "does not apply because neither field is "
-                            "ANY_SPACE.")
-                    else:
-                        info = ""
+            # 2) Check function space names
+            node1_fs_name = node1.field_space.orig_name
+            node2_fs_name = node2.field_space.orig_name
+            # 2.1) Check that both function spaces are valid
+            if not (node1_fs_name in VALID_FUNCTION_SPACE_NAMES and
+                    node2_fs_name in VALID_FUNCTION_SPACE_NAMES):
+                raise TransformationError(
+                    "Error in {0} transformation: One or both function "
+                    "spaces '{1}' and '{2}' have invalid names.".
+                    format(self.name, node1_fs_name, node2_fs_name))
+            # Check whether any of the spaces is ANY_SPACE. Loop fusion over
+            # ANY_SPACE is allowed only when the 'same_space' flag is set
```

Comment:
It may be simpler but I think it would be inaccurate and ambiguous when it comes to the `same_space` use. If I understood the original logic correctly, `same_space` was intented to be used only when at least one of the loops is on `ANY_SPACE` but the user knows they are same. Otherwise it does not make much sense.
Unconditional `same_space` flag may be dangerous if there is no additional check that the specific spaces are indeed the same (hence duplicating specific check space from the propsed "elif" condition). Also, it may be applied to two same specific spaces where it does not need to be used (as in the corrected test in `psyGen_test` where both spaces are `w1` and `same_space` is used on master.

Reply:
After a careful consideration I decided to allow the use of `same_space` flag when the function spaces are not `ANY_SPACE` but have the same name. This can be useful when loop fusion is applied generically in LFRic.